File: rotate-array/rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if k > len(nums):
            k %= len(nums)
        new_list = []
        for i in range(0, len(nums)):
            new_list.append(nums[(i + len(nums) - k) % len(nums)])
        for i in range(0, len(nums)):
            nums[i] = new_list[i]

File: rotate-array/rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n;
        
        int start = 0, count = 0;
        while (count < n) {
            int cur = start, prev = nums[start];
            while (true) {
                int next = (cur + k) % n;
                int temp = nums[next];
                nums[next] = prev;
                prev = temp;
                cur = next;
                count++;

                if (start == cur) {
                    start++;
                    break;
                }
            }

        }
    }
};
